### PR TITLE
doc: clarify syscall Z_OOPS() policy

### PR DESCRIPTION
**Pull request by tyrantlizard77**
As described in #17735.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Comment by andrewboie** [comment]
This just posts the intended policy in the documentation. There are a variety of things in the kernel that need to be fixed up to conform to it.

**Review by dcpleung — APPROVED** [review]

File: doc/reference/usermode/syscalls.rst [hunk 1]
```diff
@@ -331,6 +331,71 @@ For example:
...
+#. Some parameter checks are enforced only from user mode. These should be
+   checked in the verification function and propagate a return value to the
+   caller if possible.
+
+There are some known exceptions to these policies currently in Zephyr:
+
+* :c:func:`k_thread_join()` and :c:func:`k_thread_abort()` are no-ops if
+  the thread object isn't initialized. This is because for threads, the
+  initialization bit pulls double-duty to indicate whether a thread is
+  running, cleared upon exit. See #23020.
```
Comment (nashif) [inline doc/reference/usermode/syscalls.rst]: what is 23020? We probably need a full link, on github this look like a closed and unrelated PR.

File: doc/reference/usermode/syscalls.rst [hunk 2]
```diff
@@ -331,6 +331,71 @@ For example:
...
+#. General parameter checks should be done in the implementation function,
+   in most cases using ``CHECKIF()``.
+
+   * The behavior of ``CHECKIF()`` depends on the kernel configuration, but if
+     user mode is enabled, :option:`CONFIG_RUNTIME_ERROR_CHECKS` is enforced,
+     which guarantees that these checks will be made and a return value
+     propagated.
+
+#. It is totally forbidden for any kind of kernel mode callback function to
+   be registered from user mode. APIs which simply install callbacks should not
```
Comment (ioannisg) [inline doc/reference/usermode/syscalls.rst]: shoould -> shall, perhaps?

**Review by ceolin — COMMENTED** [review]
Looks good to me. Only comment that since we are talking about callbacks policies, memory read/write access, would be nice we expand a little bit more about when using `z_user_to_coppy` and `z_user_from_copy` or when is ok chasing a pointer.
I don't know if this is documented elsewhere though.

File: doc/reference/usermode/syscalls.rst (hunk 1 above)
Comment (andrewboie) [reply doc/reference/usermode/syscalls.rst]: I can't type, should be https://github.com/zephyrproject-rtos/zephyr/issues/23030

**Comment by andrewboie** [comment]
> 
> 
> Looks good to me. Only comment that since we are talking about callbacks policies, memory read/write access, would be nice we expand a little bit more about when using `z_user_to_coppy` and `z_user_from_copy` or when is ok chasing a pointer.
> I don't know if this is documented elsewhere though.

I'd like to punt on this, it's not directly related to Z_OOPS(). I do think some documentation illustrating what data needs to be copied back and forth from user mode, and the right way to do it, will be valuable. But I think we will learn some things when we enable virtual memory spaces in the near future and will have a better idea what to write at that time, I think we are currently making some assumptions that won't hold once we don't use a common physical memory map.

File: doc/reference/usermode/syscalls.rst (hunk 2 above)
Comment (andrewboie) [reply doc/reference/usermode/syscalls.rst]: done

**Comment by ceolin** [comment]
> > Looks good to me. Only comment that since we are talking about callbacks policies, memory read/write access, would be nice we expand a little bit more about when using `z_user_to_coppy` and `z_user_from_copy` or when is ok chasing a pointer.
> > I don't know if this is documented elsewhere though.
> 
> I'd like to punt on this, it's not directly related to Z_OOPS(). I do think some documentation illustrating what data needs to be copied back and forth from user mode, and the right way to do it, will be valuable. But I think we will learn some things when we enable virtual memory spaces in the near future and will have a better idea what to write at that time, I think we are currently making some assumptions that won't hold once we don't use a common physical memory map.

fair enough :)

**Review by ceolin — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Pull request merged**